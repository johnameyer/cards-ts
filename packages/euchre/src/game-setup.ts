import { GenericGameSetup, Intermediary } from '@cards-ts/core';
import { GameParams } from './game-params';

export class GameSetup implements GenericGameSetup<GameParams> {
    getDefaultParams(): GameParams {
        return {
            maxScore: 10,
            quickEnd: true,
        };
    }
    
    async setupForIntermediary(host: Intermediary): Promise<GameParams> {
        const [ _, resultsPromise ] = host.form(
            { type: 'input', message: [ 'Score to play to? (default 10)' ] },
            { type: 'confirm', message: [ 'End a round quickly if no points are remaining?' ] },
        );

        const results = await resultsPromise;

        const maxScore = Number(results[0]) || 100;
        const quickEnd = results[1];

        return {
            maxScore,
            quickEnd,
        };
    }

    verifyParams(params: GameParams) {
        const errors: { maxScore?: string; numToPass?: string; quickEnd?: string; } = {};
        try {
            if(params.quickEnd !== true && params.quickEnd !== false) {
                throw new Error();
            }
        } catch (e) {
            errors.quickEnd = 'Quick-end must be a boolean.';
        }
        try {
            if(!Number(params.maxScore) || Number(params.maxScore) <= 0) {
                throw new Error();
            }
        } catch (e) {
            errors.maxScore = 'Max score must be a number greater than 0';
        }
        return errors;
    }

    getYargs(): {[key: string]: import('yargs').Options} {
        return {
            quickEnd: { description: 'End a round quickly if no points are remaining', type: 'boolean', default: true },
            maxScore: { description: 'Score to play to', type: 'number', default: 10 },
        };
    }

    setupForYargs(params: any): GameParams {
        return {
            quickEnd: !!params.quickEnd,
            maxScore: Number(params.maxScore),
        };
    }
}
