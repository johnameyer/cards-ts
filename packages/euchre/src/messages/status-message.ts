import { DealOutMessage, EndRoundMessage, PlayedMessage, WonRoundMessage, NameTrumpMessage, OrderUpMessage, PassMessage, TrumpMessage, WonTrickMessage } from './status';
import { GoingAloneMessage } from './status/going-alone-message';
import { DealerMessage, LeadsMessage, TurnUpMessage } from '@cards-ts/core';

export type StatusMessage =
    DealOutMessage
    | DealerMessage
    | EndRoundMessage
    | LeadsMessage
    | PlayedMessage
    | WonRoundMessage
    | PassMessage
    | OrderUpMessage
    | NameTrumpMessage
    | TrumpMessage
    | TurnUpMessage
    | WonTrickMessage
    | GoingAloneMessage;
