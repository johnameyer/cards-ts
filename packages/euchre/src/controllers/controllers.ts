import { DefaultControllerKeys, DefaultControllers, SystemHandlerParams, Rank, ControllersProviders, DeckControllerProvider, HandsControllerProvider, PointsControllerProvider, TricksControllerProvider, TurnControllerProvider, UnwrapProviders, ValidatedProviders } from "@cards-ts/core";
import { GameParams } from "../game-params";
import { GameStates } from "../game-states";
import { ResponseMessage } from "../messages/response-message";
import { GameHandlerParams } from "../game-handler-params";
import { EuchreControllerProvider } from "./euchre-controller";

type TypedDefaultControllers = DefaultControllers<GameParams, typeof GameStates, ResponseMessage, GameHandlerParams & SystemHandlerParams>;

export const buildProviders = () => {
    const providers = {
        deck: new DeckControllerProvider(1, true, false, Rank.ranks.slice(Rank.ranks.indexOf(Rank.NINE))),
        hand: new HandsControllerProvider(),
        trick: new TricksControllerProvider(),
        score: new PointsControllerProvider(),
        euchre: new EuchreControllerProvider(),
        tricksTaken: new PointsControllerProvider(),
        turn: new TurnControllerProvider()
    };
    return providers as Omit<ValidatedProviders<typeof providers & ControllersProviders<TypedDefaultControllers>>, DefaultControllerKeys>;
};

export type GameControllers = UnwrapProviders<ReturnType<typeof buildProviders>>;

export type Controllers = GameControllers & TypedDefaultControllers;