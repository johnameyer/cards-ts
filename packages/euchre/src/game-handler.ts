import { ControllerHandlerState, DataResponseMessage, Handler, HandlerAction, PlayCardResponseMessage } from "@cards-ts/core";
import { Controllers } from "./controllers/controllers";
import { GameHandlerParams } from "./game-handler-params";
import { DealerDiscardResponseMessage, GoingAloneResponseMessage, OrderUpResponseMessage, NameTrumpResponseMessage } from "./messages/response";
import { ResponseMessage } from "./messages/response-message";

export type HandlerData = ControllerHandlerState<Controllers>;

export abstract class GameHandler implements Handler<GameHandlerParams, HandlerData, ResponseMessage> {
    abstract handleOrderUp: HandlerAction<HandlerData, DataResponseMessage | OrderUpResponseMessage | GoingAloneResponseMessage>;

    abstract handleNameTrump: HandlerAction<HandlerData, DataResponseMessage | NameTrumpResponseMessage | GoingAloneResponseMessage>;

    abstract handleDealerDiscard: HandlerAction<HandlerData, DataResponseMessage | DealerDiscardResponseMessage>;

    abstract handleTurn: HandlerAction<HandlerData, DataResponseMessage | PlayCardResponseMessage>;
}