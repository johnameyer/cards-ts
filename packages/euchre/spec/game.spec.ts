import { expect } from 'chai';
import { IntermediaryHandler } from '../src/index.js';
import { eventHandler } from '../src/event-handler.js';
import { GameSetup } from '../src/game-setup.js';
import { gameStateTransitions } from '../src/game-state-transitions.js';
import { GameParams } from '../src/game-params.js';
import { buildProviders } from '../src/controllers/controllers.js';
import { HeuristicHandler } from '../src/handlers/heuristic-handler.js';
import { ArrayMessageHandler, buildGameFactory, DeckControllerProvider, HandlerChain, Rank } from '@cards-ts/core';
import { GameHandler } from '../src/game-handler.js';
import { StatusMessage } from '../../can-i-have-that/src/messages/status-message.js';

describe('game', () => {
    // TODO can we build this more simply i.e. deterministic deck controller?
    const mockProviders = () => ({
        ...buildProviders(),
        // TODO how to nicely have deterministic order by default in tests?
        deck: new DeckControllerProvider(1, false, false, Rank.ranks.slice(Rank.ranks.indexOf(Rank.NINE))), // note shouldShuffle false
    });

    const factory = buildGameFactory(
        gameStateTransitions,
        eventHandler,
        // TODO can we avoid specifying the setup and intermediary since they are not critical to the driver construction?
        new GameSetup(),
        intermediary => new IntermediaryHandler(intermediary),
        () => new HeuristicHandler(),
        mockProviders,
    );

    it('works as expected', async () => {
        const messageHandlers = Array.from({length: 4}, () => new ArrayMessageHandler<StatusMessage>());

        const gameHandler: () => GameHandler = () => new HeuristicHandler();

        // @ts-expect-error
        const players = Array.from(messageHandlers, messageHandler => new HandlerChain([ messageHandler, gameHandler() ]));

        const params: GameParams = {
            ...new GameSetup().getDefaultParams(),
            maxScore: 4,
        };

        const names = Array.from(messageHandlers, (_, i) => String.fromCharCode(65 + i));

        const driver = factory.getGameDriver(players, params, names);

        driver.resume();

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr[0].type).to.equal('deal-out-message');

            messageHandler.clear();
        }

        while(driver.getState().euchre.bidder === undefined) {
            driver.handleSyncResponses();
            driver.resume();
        }

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr[0].type).to.equal('pass-message');
            expect(messageHandler.arr[1].type).to.equal('pass-message');
            expect(messageHandler.arr[2].type).to.equal('pass-message');
            expect(messageHandler.arr[3].type).to.equal('order-up-message');
            expect(messageHandler.arr[4].type).to.equal('trump-message');

            messageHandler.clear();
        }

        // TODO test further
    });
});
