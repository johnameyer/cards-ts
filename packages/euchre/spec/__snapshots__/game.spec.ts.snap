// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game has a matching snapshot 1`] = `
Array [
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "C played J of Diamonds",
  "D played Q of Diamonds",
  "A took the trick",
  "A leads next trick",
  "B played A of Diamonds",
  "C played 9 of Spades",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "B played 10 of Hearts",
  "C played J of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "B played Q of Clubs",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "C played K of Clubs",
  "D played A of Clubs",
  "A took the trick",
  "A, C won 1 points",
  "A: 1 B: 0 C: 1 D: 0",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "D played J of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "C played A of Diamonds",
  "D played 9 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "A took the trick",
  "A leads next trick",
  "B played K of Hearts",
  "C played Q of Clubs",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "D played K of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 1 B: 1 C: 1 D: 1",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "B played Q of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "D played A of Diamonds",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "D played 10 of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played K of Hearts",
  "D played Q of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "B played A of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 2 B: 1 C: 2 D: 1",
  "Received ",
  "J of Spades was turned over",
  "A passed",
  "B passed",
  "C passed",
  "D ordered up the card",
  "Spades is trump",
  "A leads next trick",
  "B played J of Diamonds",
  "C played Q of Diamonds",
  "D played K of Diamonds",
  "D took the trick",
  "D leads next trick",
  "D played J of Clubs",
  "B played 9 of Spades",
  "C played 10 of Spades",
  "D took the trick",
  "D leads next trick",
  "D played 9 of Hearts",
  "B played J of Hearts",
  "C played Q of Hearts",
  "C took the trick",
  "C leads next trick",
  "C played 10 of Clubs",
  "D played K of Hearts",
  "B played 9 of Clubs",
  "A took the trick",
  "A leads next trick",
  "B played K of Clubs",
  "C played A of Clubs",
  "D played J of Spades",
  "D took the trick",
  "B, D won 1 points",
  "A: 2 B: 2 C: 2 D: 2",
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "C played J of Diamonds",
  "D played Q of Diamonds",
  "A took the trick",
  "A leads next trick",
  "B played A of Diamonds",
  "C played 9 of Spades",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "B played 10 of Hearts",
  "C played J of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "B played Q of Clubs",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "C played K of Clubs",
  "D played A of Clubs",
  "A took the trick",
  "A, C won 1 points",
  "A: 3 B: 2 C: 3 D: 2",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "D played J of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "C played A of Diamonds",
  "D played 9 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "A took the trick",
  "A leads next trick",
  "B played K of Hearts",
  "C played Q of Clubs",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "D played K of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 3 B: 3 C: 3 D: 3",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "B played Q of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "D played A of Diamonds",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "D played 10 of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played K of Hearts",
  "D played Q of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "B played A of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 4 B: 3 C: 4 D: 3",
]
`;

exports[`game has a matching snapshot 2`] = `
Array [
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "C played J of Diamonds",
  "D played Q of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "C played 9 of Spades",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "C played J of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "A played K of Hearts",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "C played K of Clubs",
  "D played A of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 1 B: 0 C: 1 D: 0",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "D played J of Diamonds",
  "A played Q of Diamonds",
  "B took the trick",
  "B leads next trick",
  "C played A of Diamonds",
  "D played 9 of Spades",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "C played Q of Clubs",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "D played K of Clubs",
  "A played A of Clubs",
  "B took the trick",
  "B, D won 1 points",
  "A: 1 B: 1 C: 1 D: 1",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "A played J of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "D played A of Diamonds",
  "A played 9 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "D played 10 of Hearts",
  "A played J of Hearts",
  "B took the trick",
  "B leads next trick",
  "C played K of Hearts",
  "D played Q of Clubs",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "A played K of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 2 B: 1 C: 2 D: 1",
  "Received ",
  "J of Spades was turned over",
  "A passed",
  "B passed",
  "C passed",
  "D ordered up the card",
  "Spades is trump",
  "A leads next trick",
  "A played 10 of Diamonds",
  "C played Q of Diamonds",
  "D played K of Diamonds",
  "D took the trick",
  "D leads next trick",
  "D played J of Clubs",
  "A played A of Diamonds",
  "C played 10 of Spades",
  "D took the trick",
  "D leads next trick",
  "D played 9 of Hearts",
  "A played 10 of Hearts",
  "C played Q of Hearts",
  "C took the trick",
  "C leads next trick",
  "C played 10 of Clubs",
  "D played K of Hearts",
  "A played Q of Clubs",
  "A took the trick",
  "A leads next trick",
  "A played A of Hearts",
  "C played A of Clubs",
  "D played J of Spades",
  "D took the trick",
  "B, D won 1 points",
  "A: 2 B: 2 C: 2 D: 2",
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "C played J of Diamonds",
  "D played Q of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "C played 9 of Spades",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "C played J of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "A played K of Hearts",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "C played K of Clubs",
  "D played A of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 3 B: 2 C: 3 D: 2",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "D played J of Diamonds",
  "A played Q of Diamonds",
  "B took the trick",
  "B leads next trick",
  "C played A of Diamonds",
  "D played 9 of Spades",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "C played Q of Clubs",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "D played K of Clubs",
  "A played A of Clubs",
  "B took the trick",
  "B, D won 1 points",
  "A: 3 B: 3 C: 3 D: 3",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "A played J of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "D played A of Diamonds",
  "A played 9 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "D played 10 of Hearts",
  "A played J of Hearts",
  "B took the trick",
  "B leads next trick",
  "C played K of Hearts",
  "D played Q of Clubs",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "A played K of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 4 B: 3 C: 4 D: 3",
]
`;

exports[`game has a matching snapshot 3`] = `
Array [
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "D played Q of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "B played A of Diamonds",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "B played 10 of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "A played K of Hearts",
  "B played Q of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "D played A of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 1 B: 0 C: 1 D: 0",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "D played J of Diamonds",
  "A played Q of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "D played 9 of Spades",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "D played J of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "B played K of Hearts",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "D played K of Clubs",
  "A played A of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 1 B: 1 C: 1 D: 1",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "A played J of Diamonds",
  "B played Q of Diamonds",
  "C took the trick",
  "C leads next trick",
  "D played A of Diamonds",
  "A played 9 of Spades",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "D played 10 of Hearts",
  "A played J of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "D played Q of Clubs",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "A played K of Clubs",
  "B played A of Clubs",
  "C took the trick",
  "A, C won 1 points",
  "A: 2 B: 1 C: 2 D: 1",
  "Received ",
  "J of Spades was turned over",
  "A passed",
  "B passed",
  "C passed",
  "D ordered up the card",
  "Spades is trump",
  "A leads next trick",
  "A played 10 of Diamonds",
  "B played J of Diamonds",
  "D played K of Diamonds",
  "D took the trick",
  "D leads next trick",
  "D played J of Clubs",
  "A played A of Diamonds",
  "B played 9 of Spades",
  "D took the trick",
  "D leads next trick",
  "D played 9 of Hearts",
  "A played 10 of Hearts",
  "B played J of Hearts",
  "C took the trick",
  "C leads next trick",
  "D played K of Hearts",
  "A played Q of Clubs",
  "B played 9 of Clubs",
  "A took the trick",
  "A leads next trick",
  "A played A of Hearts",
  "B played K of Clubs",
  "D played J of Spades",
  "D took the trick",
  "B, D won 1 points",
  "A: 2 B: 2 C: 2 D: 2",
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "D played Q of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "B played A of Diamonds",
  "D played 10 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "B played 10 of Hearts",
  "D played Q of Hearts",
  "D took the trick",
  "D leads next trick",
  "D played 10 of Clubs",
  "A played K of Hearts",
  "B played Q of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "D played A of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 3 B: 2 C: 3 D: 2",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "D played J of Diamonds",
  "A played Q of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "D played 9 of Spades",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "D played J of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "B played K of Hearts",
  "D played 9 of Clubs",
  "C took the trick",
  "C leads next trick",
  "D played K of Clubs",
  "A played A of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 3 B: 3 C: 3 D: 3",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "D played 10 of Diamonds",
  "A played J of Diamonds",
  "B played Q of Diamonds",
  "C took the trick",
  "C leads next trick",
  "D played A of Diamonds",
  "A played 9 of Spades",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "D played 10 of Hearts",
  "A played J of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "D played Q of Clubs",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "D played A of Hearts",
  "A played K of Clubs",
  "B played A of Clubs",
  "C took the trick",
  "A, C won 1 points",
  "A: 4 B: 3 C: 4 D: 3",
]
`;

exports[`game has a matching snapshot 4`] = `
Array [
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "C played J of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "B played A of Diamonds",
  "C played 9 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "B played 10 of Hearts",
  "C played J of Hearts",
  "D took the trick",
  "D leads next trick",
  "A played K of Hearts",
  "B played Q of Clubs",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "C played K of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 1 B: 0 C: 1 D: 0",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "A played Q of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "C played A of Diamonds",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "C played 10 of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "B played K of Hearts",
  "C played Q of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "A played A of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 1 B: 1 C: 1 D: 1",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "A played J of Diamonds",
  "B played Q of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "A played 9 of Spades",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "A played J of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played K of Hearts",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "A played K of Clubs",
  "B played A of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 2 B: 1 C: 2 D: 1",
  "Received ",
  "J of Spades was turned over",
  "A passed",
  "B passed",
  "C passed",
  "D ordered up the card",
  "Spades is trump",
  "A leads next trick",
  "A played 10 of Diamonds",
  "B played J of Diamonds",
  "C played Q of Diamonds",
  "D took the trick",
  "D leads next trick",
  "A played A of Diamonds",
  "B played 9 of Spades",
  "C played 10 of Spades",
  "D took the trick",
  "D leads next trick",
  "A played 10 of Hearts",
  "B played J of Hearts",
  "C played Q of Hearts",
  "C took the trick",
  "C leads next trick",
  "C played 10 of Clubs",
  "A played Q of Clubs",
  "B played 9 of Clubs",
  "A took the trick",
  "A leads next trick",
  "A played A of Hearts",
  "B played K of Clubs",
  "C played A of Clubs",
  "D took the trick",
  "B, D won 1 points",
  "A: 2 B: 2 C: 2 D: 2",
  "Received ",
  "J of Spades was turned over",
  "B passed",
  "C passed",
  "D passed",
  "A ordered up the card",
  "Spades is trump",
  "B leads next trick",
  "B played 10 of Diamonds",
  "C played J of Diamonds",
  "A played K of Diamonds",
  "A took the trick",
  "A leads next trick",
  "A played J of Clubs",
  "B played A of Diamonds",
  "C played 9 of Spades",
  "A took the trick",
  "A leads next trick",
  "A played 9 of Hearts",
  "B played 10 of Hearts",
  "C played J of Hearts",
  "D took the trick",
  "D leads next trick",
  "A played K of Hearts",
  "B played Q of Clubs",
  "C played 9 of Clubs",
  "B took the trick",
  "B leads next trick",
  "B played A of Hearts",
  "C played K of Clubs",
  "A played J of Spades",
  "A took the trick",
  "A, C won 1 points",
  "A: 3 B: 2 C: 3 D: 2",
  "Received ",
  "J of Spades was turned over",
  "C passed",
  "D passed",
  "A passed",
  "B ordered up the card",
  "Spades is trump",
  "C leads next trick",
  "C played 10 of Diamonds",
  "A played Q of Diamonds",
  "B played K of Diamonds",
  "B took the trick",
  "B leads next trick",
  "B played J of Clubs",
  "C played A of Diamonds",
  "A played 10 of Spades",
  "B took the trick",
  "B leads next trick",
  "B played 9 of Hearts",
  "C played 10 of Hearts",
  "A played Q of Hearts",
  "A took the trick",
  "A leads next trick",
  "A played 10 of Clubs",
  "B played K of Hearts",
  "C played Q of Clubs",
  "C took the trick",
  "C leads next trick",
  "C played A of Hearts",
  "A played A of Clubs",
  "B played J of Spades",
  "B took the trick",
  "B, D won 1 points",
  "A: 3 B: 3 C: 3 D: 3",
  "Received ",
  "J of Spades was turned over",
  "D passed",
  "A passed",
  "B passed",
  "C ordered up the card",
  "Spades is trump",
  "D leads next trick",
  "A played J of Diamonds",
  "B played Q of Diamonds",
  "C played K of Diamonds",
  "C took the trick",
  "C leads next trick",
  "C played J of Clubs",
  "A played 9 of Spades",
  "B played 10 of Spades",
  "C took the trick",
  "C leads next trick",
  "C played 9 of Hearts",
  "A played J of Hearts",
  "B played Q of Hearts",
  "B took the trick",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played K of Hearts",
  "A played 9 of Clubs",
  "D took the trick",
  "D leads next trick",
  "A played K of Clubs",
  "B played A of Clubs",
  "C played J of Spades",
  "C took the trick",
  "A, C won 1 points",
  "A: 4 B: 3 C: 4 D: 3",
]
`;
